### PR TITLE
Optional exception

### PR DESCRIPTION
**Pull request by JohnLangford**
Tweaked #1897

### CODE REVIEW

**Comment by jackgerrits** [comment]
I do not think rethrowing with an exception pointer is the idiomatic way to do this

File: vowpalwabbit/vw_exception.h [hunk 1]
```diff
@@ -105,6 +105,15 @@ class vw_unrecognised_option_exception : public vw_exception
   ~vw_unrecognised_option_exception() _NOEXCEPT {}
 };
 
+class strict_parse_exception : public std::exception
```
Comment (jackgerrits) [inline vowpalwabbit/vw_exception.h]: Should inherit from vw_exception

Comment (JohnLangford) [reply vowpalwabbit/vw_exception.h]: That makes some sense.

**Comment by JohnLangford** [comment]
Can you use the command line instructions to create your own variation on this patch?  (I'm busy with kids right now.)

**Comment by jackgerrits** [comment]
Yeah, sure thing

**Comment by jackgerrits** [comment]
So I made the changes, but the print statement used for the vw_exception prints the file and line number of the exception. This changes depending on filesystem on every machine. Not sure how to resolve that here.

vw (/mnt/c/w/repos/vowpal_wabbit_jackgerrits/vowpalwabbit/parse_example.cc:85): malformed example! '|',space, or EOL expected after : "| x:0.7"in Example #0: "| x:0.7"

**Comment by JohnLangford** [comment]
See comment here: https://github.com/jackgerrits/vowpal_wabbit/commit/f42678d4e57ddcdccaba7cacc5001cb920b89d12#diff-5e0723881d65ba028f55a90faa0f6d93R112

Also note: it looks like you commented out example dealloc -> create memory leak.

**Comment by jackgerrits** [comment]
I just realized changing the negation script was wrong -- please wait to merge

File: test/negative-test.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+# this script runs a command and ensures that its exit code is not zero.
+
+if $@; then
+    exit 0
```
Comment (jackgerrits) [inline test/negative-test.sh]:
```suggestion
    exit 1
```

File: test/negative-test.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+else
+    exit 1
```
Comment (jackgerrits) [inline test/negative-test.sh]:
```suggestion
    exit 0
```

**Comment by JohnLangford** [comment]
Thanks @palmerlao @jackgerrits @rajan-chari   

@rajan-chari opening an issue around how we handle errors in parsers seems like a good idea.

**Pull request merged**